system_zero/src/lookup.rs

```diff
@@ -0,0 +1,114 @@
+//! Implementation of the Halo2 lookup argument.
+//!
+//! References:
+//! - https://zcash.github.io/halo2/design/proving-system/lookup.html
+//! - https://www.youtube.com/watch?v=YlTt12s7vGE&t=5237s
+
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::{Field, PrimeField64};
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use starky::vars::StarkEvaluationTargets;
+use starky::vars::StarkEvaluationVars;
+
+use crate::public_input_layout::NUM_PUBLIC_INPUTS;
+use crate::registers::lookup::*;
+use crate::registers::NUM_COLUMNS;
+use crate::util::{create_hash_bag, flatten_hash_bag};
+
+pub(crate) fn generate_lookups<F: PrimeField64>(trace_cols: &mut [Vec<F>]) {
+    for i in 0..NUM_LOOKUPS {
+        let inputs = &trace_cols[col_input(i)];
+        let table = &trace_cols[col_table(i)];
+        let (permuted_inputs, permuted_table) = permuted_cols(inputs, table);
+        trace_cols[col_permuted_input(i)] = permuted_inputs;
+        trace_cols[col_permuted_table(i)] = permuted_table;
+    }
+}
+
+/// Given an input column and a table column, generate the permuted input and permuted table columns
+/// used in the Halo2 permutation argument.
+pub(crate) fn permuted_cols<F: PrimeField64>(inputs: &[F], table: &[F]) -> (Vec<F>, Vec<F>) {
+    let n = inputs.len();
+
+    // In the permuted inputs, copies of the same value must be grouped together. We accomplish this
+    // by building a hash-based bag (aka multiset) from the inputs, then flattening it.
+    let input_bag = create_hash_bag(inputs);
+    let permuted_inputs = flatten_hash_bag(&input_bag);
+
+    // Enumerate tables values that do not appear in the input list.
+    let mut unused_table_vals = table.iter().filter(|v| !input_bag.contains_key(v)).copied();
```

Comment:
I think this is wrong when `table` has duplicate elements. For example, if `table = [0; n]`, `unused_table_vals` will be empty and we'll get a panic in the `next().expect()` below.

A working version would be:
```rust
let mut unused_table_vals = create_hash_bag(&table);
for &x in input_bag.keys() {
    unused_table_vals.entry(x).and_modify(|v| *v -= 1);
}
```

Reply:
Fixed by deleting that method :)